acknowledger.go

```diff
@@ -1,37 +1,136 @@
 package amqprpc
 
 import (
+	"context"
+	"errors"
+	"log/slog"
+
 	amqp "github.com/rabbitmq/amqp091-go"
 )
 
-// AwareAcknowledger implements the amqp.Acknowledger interface with the
-// addition that it can tell if a message has been acked in any way.
-type AwareAcknowledger struct {
+// SafeAcknowledgeHandler implements the amqp.Acknowledger interface with the
+// addition that it can tell if a message has been acked in any way, and
+// if the ack action fails - it will close the underlying connection.
+type SafeAcknowledgeHandler struct {
 	Acknowledger amqp.Acknowledger
 	Handled      bool
+	conn         *amqp.Channel
 }
 
-// NewAwareAcknowledger returns the passed acknowledger as an AwareAcknowledger.
-func NewAwareAcknowledger(acknowledger amqp.Acknowledger) *AwareAcknowledger {
-	return &AwareAcknowledger{
+// NewSafeAcknowledgeHandler returns the passed acknowledger as an AwareAcknowledger.
+func NewSafeAcknowledgeHandler(acknowledger amqp.Acknowledger, ch *amqp.Channel) *SafeAcknowledgeHandler {
+	return &SafeAcknowledgeHandler{
 		Acknowledger: acknowledger,
+		conn:         ch,
 	}
 }
 
 // Ack passes the Ack down to the underlying Acknowledger.
-func (a *AwareAcknowledger) Ack(tag uint64, multiple bool) error {
+// Will close the underlying connection if ack fails.
+func (a *SafeAcknowledgeHandler) Ack(tag uint64, multiple bool) error {
+	if a.Handled {
```

Comment:
Since Ack can be called by multiple goroutines we actually have a race here. Perhaps a `Handled atomic.Bool` is better?

```
	if a.Handled.CompareAndSwap(false, true) {
		return a.Acknowledger.Ack(tag, multiple)
	}

	return nil
```